pumpkin-protocol/src/client/play/c_commands.rs

```diff
@@ -0,0 +1,382 @@
+        is_executable: bool,
+    },
+    Argument {
+        name: &'a str,
+        is_executable: bool,
+        parser: ProtoCmdArgParser<'a>,
+        override_suggestion_type: Option<ProtoCmdArgSuggestionType>,
+    },
+}
+
+impl<'a> ProtoNode<'a> {
+    const FLAG_IS_EXECUTABLE: i8 = 4;
+    const FLAG_HAS_REDIRECT: i8 = 8;
+    const FLAG_HAS_SUGGESTION_TYPE: i8 = 16;
+
+    /// https://wiki.vg/Command_Data
+    pub fn write_to(&self, bytebuf: &mut ByteBuffer) {
+        // flags
+        let flags = match self.node_type {
+            ProtoNodeType::Root => 0,
+            ProtoNodeType::Literal {
+                name: _,
+                is_executable,
+            } => {
+                let mut n = 1;
+                if is_executable {
+                    n |= Self::FLAG_IS_EXECUTABLE
+                }
+                n
+            }
+            ProtoNodeType::Argument {
+                name: _,
+                is_executable,
+                parser: _,
+                override_suggestion_type: override_suggestion_tpye,
+            } => {
+                let mut n = 2;
+                if override_suggestion_tpye.is_some() {
+                    n |= Self::FLAG_HAS_SUGGESTION_TYPE
+                }
+                if is_executable {
+                    n |= Self::FLAG_IS_EXECUTABLE
+                }
+                n
+            }
+        };
+        bytebuf.put_i8(flags);
+
+        // child count + children
+        bytebuf.put_list(&self.children, |bytebuf, child| bytebuf.put_var_int(child));
+
+        // redirect node
+        if flags & Self::FLAG_HAS_REDIRECT != 0 {
+            bytebuf.put_var_int(&1.into());
+        }
+
+        // name
+        match self.node_type {
+            ProtoNodeType::Argument { name, .. } | ProtoNodeType::Literal { name, .. } => {
+                bytebuf.put_string_len(name, 32767)
+            }
+            ProtoNodeType::Root => {}
+        }
+
+        // parser id + properties
+        if let ProtoNodeType::Argument {
+            name: _,
+            is_executable: _,
+            parser,
+            override_suggestion_type: _,
+        } = &self.node_type
+        {
+            parser.write_to_buffer(bytebuf)
+        }
+
+        if flags & Self::FLAG_HAS_SUGGESTION_TYPE != 0 {
+            match &self.node_type {
+                ProtoNodeType::Argument { name: _, is_executable: _, parser: _, override_suggestion_type } => {
+                    // suggestion type
+                    let suggestion_type = &override_suggestion_type.expect("ProtoNode::FLAG_HAS_SUGGESTION_TYPE should only be set if override_suggestion_type is not None.");
+                    bytebuf.put_string(suggestion_type.identifier());
+                },
+                _ => unimplemented!("ProtoNode::FLAG_HAS_SUGGESTION_TYPE is only implemented for ProtoNodeType::Argument"),
+            }
+        }
+    }
+}
+
+/// see https://wiki.vg/Command_Data for meaning of flags/enums/etc.
```

Comment:
I recommend against using wiki.vg

Reply:
It was pretty helpful actually, but as I said above, I for some reason limited myself to text search when browsing vanilla source code so it was a struggle finding the relevant spots. I'll remove that comment